### PR TITLE
[FORK][x64] Remove custom binary PReLU post op

### PR DESCRIPTION
**Pull request by maxnick**
# Description

OV PR: https://github.com/openvinotoolkit/openvino/pull/23963

### CODE REVIEW

File: src/cpu/primitive_attr_postops.cpp [hunk 1]
```diff
@@ -184,8 +183,6 @@ ref_post_ops_t::ref_post_ops_t(const post_ops_t &po, bool skip_sum)
             eltwise_po_.emplace_back(e.eltwise);
         } else if (po_.contain(primitive_kind::binary, idx)) {
             binary_po_.emplace_back(e.binary);
-        } else if (po_.contain(primitive_kind::depthwise, idx)) {
```
Comment (luweizhou2016) [inline src/cpu/primitive_attr_postops.cpp]: Why removing this post ops? ref pp doesn't use? only jit_pp would use?

File: src/cpu/x64/injectors/jit_uni_binary_injector.cpp [hunk 2]
```diff
@@ -2596,32 +2592,6 @@ jit_uni_binary_injector_t<isa, Vmm>::execute_cmp_binary(const Vmm &dst,
...
-    push_vmm(host_, zmm_aux0);
-
-    host_->uni_vpxor(zmm_aux0, zmm_aux0, zmm_aux0);
-    host_->vcmpps(aux_kmask | maybe_tail_kmask, lhs, zmm_aux0, jit_generator::_cmp_lt_os);
-    pop_vmm(host_, zmm_aux0);
-    host_->uni_vmulps(dst_vmm | aux_kmask, lhs, rhs);
-    pop_opmask(host_, aux_kmask);
-}
-
-
```
Comment (luweizhou2016) [inline src/cpu/x64/injectors/jit_uni_binary_injector.cpp:2624]: My understanding is the that after this PR, the jit_uni_binary_injector.cpp/hpp would be the same with onednn upstream. Right?

File: src/cpu/x64/jit_gemm_inner_product_utils.cpp [hunk 3]
```diff
@@ -334,7 +333,7 @@ jit_pp_kernel_t<isa>::jit_pp_kernel_t(size_t OC, size_t MB, dim_t dst_mb_stride,
                 helper_vmm_idx, eltwise_reserved_gpr_, r14, r15, preserve_gpr,
                 preserve_vmm, PARAM_OFF(post_ops_binary_rhs_arg_vec),
                 PARAM_OFF(dst_orig), dst_md_wrapper, tail_size, opmask_binary,
-                reg_tmp, use_exact_tail_scalar_bcast, prelu_helper_vmm_idx};
+                reg_tmp, use_exact_tail_scalar_bcast};
         static const bcast_set_t enabled_bcast_strategy
                 = {broadcasting_strategy_t::scalar,
                         broadcasting_strategy_t::per_oc,
```
Comment (luweizhou2016) [inline src/cpu/x64/jit_gemm_inner_product_utils.cpp:339]: So this file would be the same with onednn upstream?

**Review by luweizhou2016 — COMMENTED** [review]
The gemm conv kernel diverges most with the upstream version, because fork enabled jit_pp kernel for jit gemm kernel but upstream still use ref_pp kernel. 
https://github.com/openvinotoolkit/oneDNN/blob/26633ae49edd4353a29b7170d9fcef6b2d79f4b3/src/cpu/gemm_convolution.hpp#L77 shows that gemm_conv should not support prelu after this change.  
I just checked that we have test cases of checking prelu fusing on jit gemm kernel, but I didn't see these test cases are disabled on OV pr. can you pls help to double check? 
I think we need to add the prelu into the list. Some older platforms maybe would fallback on gemm conv on some cases.

File: src/cpu/x64/injectors/jit_uni_binary_injector.cpp (hunk 2 above)
Comment (maxnick) [reply src/cpu/x64/injectors/jit_uni_binary_injector.cpp:2624]: The intension was to remove all the custom binary prelu related code, whether there is some extra OV specific code is still unknown.

File: src/cpu/x64/jit_gemm_inner_product_utils.cpp (hunk 3 above)
Comment (maxnick) [reply src/cpu/x64/jit_gemm_inner_product_utils.cpp:339]: The intension was to remove all the custom binary prelu related code, whether there is some extra OV specific code is still unknown.

**Comment by maxnick** [comment]
> The gemm conv kernel diverges most with the upstream version, because fork enabled jit_pp kernel for jit gemm kernel but upstream still use ref_pp kernel.
> 
> https://github.com/openvinotoolkit/oneDNN/blob/26633ae49edd4353a29b7170d9fcef6b2d79f4b3/src/cpu/gemm_convolution.hpp#L77
> 
> shows that gemm_conv should not support prelu after this change.
> I just checked that we have test cases of checking prelu fusing on jit gemm kernel, but I didn't see these test cases are disabled on OV pr. can you pls help to double check?
> I think we need to add the prelu into the list. Some older platforms maybe would fallback on gemm conv on some cases.

Thank you for bringing attention to this. Need to check. Do I correctly understand that jit_pp kernel use the oneDNN post ops as is, and all we need is to simply add prelu to this list and the oneDNN prelu post op will work as it is within this jit_pp kernel?

**Comment by luweizhou2016** [comment]
> > The gemm conv kernel diverges most with the upstream version, because fork enabled jit_pp kernel for jit gemm kernel but upstream still use ref_pp kernel.
> > https://github.com/openvinotoolkit/oneDNN/blob/26633ae49edd4353a29b7170d9fcef6b2d79f4b3/src/cpu/gemm_convolution.hpp#L77
> > 
> > shows that gemm_conv should not support prelu after this change.
> > I just checked that we have test cases of checking prelu fusing on jit gemm kernel, but I didn't see these test cases are disabled on OV pr. can you pls help to double check?
> > I think we need to add the prelu into the list. Some older platforms maybe would fallback on gemm conv on some cases.
> 
> Thank you for bringing attention to this. Need to check. Do I correctly understand that jit_pp kernel use the oneDNN post ops as is, and all we need is to simply add prelu to this list and the oneDNN prelu post op will work as it is within this jit_pp kernel?

Not sure but it should be.  because finally jit pp kernel also use jit_uni_binary_injector to do the post-ops, maybe needs to change some passing argument to binary injector.

**Comment by maxnick** [comment]
> > > The gemm conv kernel diverges most with the upstream version, because fork enabled jit_pp kernel for jit gemm kernel but upstream still use ref_pp kernel.
> > > https://github.com/openvinotoolkit/oneDNN/blob/26633ae49edd4353a29b7170d9fcef6b2d79f4b3/src/cpu/gemm_convolution.hpp#L77
> > > 
> > > shows that gemm_conv should not support prelu after this change.
> > > I just checked that we have test cases of checking prelu fusing on jit gemm kernel, but I didn't see these test cases are disabled on OV pr. can you pls help to double check?
> > > I think we need to add the prelu into the list. Some older platforms maybe would fallback on gemm conv on some cases.
> > 
> > 
> > Thank you for bringing attention to this. Need to check. Do I correctly understand that jit_pp kernel use the oneDNN post ops as is, and all we need is to simply add prelu to this list and the oneDNN prelu post op will work as it is within this jit_pp kernel?
> 
> Not sure but it should be. because finally jit pp kernel also use jit_uni_binary_injector to do the post-ops, maybe needs to change some passing argument to binary injector.

1. There were no PReLU per channel SL tests, so all the existing PReLU post ops tests of conv layers were per tensor, and the prelu post op was eventually mapped to the eltwise post op.
2. I extended the conv SL tests with PReLU per channel, but normally those post ops are performed via `dnnl_depthwise_prelu`, i.e. the legacy code path. So I brought back the depthwise post op check to the reference post ops implementation, to support prelu being performed via the legacy depthwise post op for reference convolution implementations.
3. I extended jit_pp_kernel with all the necessary checks for the original oneDNN prelu post op. I tested this code path with SL tests with PReLU per  channel post ops but with a custom build, where legacy post ops are disabled. The jit_pp_kernel works with the original oneDNN prelu post op. At least I didn't  observe any functional issues.

Hope this answers your initial question

File: src/cpu/primitive_attr_postops.cpp (hunk 1 above)
Comment (maxnick) [reply src/cpu/primitive_attr_postops.cpp]: I brought this check back.

**Review by luweizhou2016 — APPROVED** [review]
LTGM! Thx!

**Pull request closed without merging**